In [5]:
import pandas as pd

# Работа с Pandas

1. Загрузите ваш DataFrame в переменную df и выведете на экран первые пять строк 

In [6]:
df = pd.read_csv('googleplaystore .csv')
df.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type  Price Content Rating  \
0     159   19M      10,000+  Free    0.0       Everyone   
1     967   14M     500,000+  Free    0.0       Everyone   
2   87510  8.7M   5,000,000+  Free    0.0       Everyone   
3  215644   25M  50,000,000+  Free    0.0           Teen   
4     967  2.8M     100,000+  Free    0.0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

2. Отобразите количество строк и столбцов в DataFrame

In [7]:
df.shape

(10841, 13)

3. Выведете основную информацию о DataFrame методом info(). Проанализируйте есть ли пропуски в столбцах. Напишите вывод о том, какие столбцы содержат пропуски

In [8]:
df.info() 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  float64
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(2), object(11)
memory usage: 1.1+ MB


<b>Выводы</b>

* в датафрейме 13 столбцов
* 11 из них имеют тип `object` (аналогия `string` из Python)
* 2 из них имеют тип `float64`
* 0 столбцов имеют тип `int64`
* всего в датафрейме 10841 строк
* в столбцах `Rating`, `Type`, `Content Rating`, `Current Ver` и `Android Ver` есть пропуски (`NaN`)
* объем занимаемый датафреймом 1.1+ MB (эта информация может быть полезна, чтобы заранее представить какие операции с датафреймом вы сможете выполнять с учетом доступной вам оперативной памяти)

In [9]:
column_names = list(df.columns) # Функция list создаёт список из значений, которые передали внутрь. Работа аналогична обычным квадратным скобкам []
print("Название колонок:", column_names)
print("Количество колонок:", len(column_names))

Название колонок: ['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']
Количество колонок: 13


In [10]:
# Напишем просто цикл, Чтобы привести все названия столбцов к нижнему регистру

lower_column_names = [] # Создадим пустой список куда будем добавлять наши значения

for word in column_names: # Войдём в цикл по списку column_names
    lower_column_names.append(word.lower()) # Добавляем в список lower_column_names имя колонки в нижнем регистре используя метод .lower()

print(lower_column_names)

df.columns = lower_column_names # Заменяем названия столбцов на lower_column_names
df.head(3)

['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type', 'price', 'content rating', 'genres', 'last updated', 'current ver', 'android ver']


app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   

  reviews  size    installs  type  price content rating  \
0     159   19M     10,000+  Free    0.0       Everyone   
1     967   14M    500,000+  Free    0.0       Everyone   
2   87510  8.7M  5,000,000+  Free    0.0       Everyone   

                      genres      last updated current ver   android ver  
0               Art & Design   January 7, 2018       1.0.0  4.0.3 and up  
1  Art & Design;Pretend Play  January 15, 2018       2.0.0  4.0.3 and up  
2               Art & Design    August 1, 2018       1.2.4  4.0.3 and up

In [11]:
df = df.rename(columns={'content rating': 'content_rating'})
df = df.rename(columns={'last updated': 'last_updated'})
df = df.rename(columns={'current ver': 'current_ver'})
df = df.rename(columns={'android ver': 'android_ver'})

4. Удалите или замените пропуски известными вам методами. (Подсказка: `dropna()` или `fillna()`, выбор того или иного метода должен быть аргументирован).

In [12]:
df.isnull().sum()

app                  0
category             0
rating            1474
reviews              0
size                 0
installs             0
type                 1
price                0
content_rating       1
genres               0
last_updated         0
current_ver          8
android_ver          3
dtype: int64

In [13]:
df.rating.fillna(0, inplace=True) # Заменяем значение на 0 тк нет рейтинга для 1474 приложений

C:\Users\olyao\AppData\Local\Temp\ipykernel_16828\4037663921.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df.rating.fillna(0, inplace=True) # Заменяем значение на 0 тк нет рейтинга для 1474 приложений


In [14]:
df.type.value_counts() # Видим пропуск

type
Free    10039
Paid      800
0           1
Name: count, dtype: int64

In [15]:
df.content_rating.value_counts() # Видим пропуски

content_rating
Everyone           8714
Teen               1208
Mature 17+          499
Everyone 10+        414
Adults only 18+       3
Unrated               2
Name: count, dtype: int64

In [16]:
df.dropna(inplace=True) # Видим что пропусков осталось 14 шт из всего датасэта и можем их стереть

5. Проверьте нет ли пропусков после произведенной обработки

In [17]:
df.isnull().sum() 

app               0
category          0
rating            0
reviews           0
size              0
installs          0
type              0
price             0
content_rating    0
genres            0
last_updated      0
current_ver       0
android_ver       0
dtype: int64

6. Выведите на экран минимум и максимум из столбца `Price`:

In [18]:
df['price'].min()

0.0

In [19]:
df['price'].max()

400.0

7. Выведите на экран медиану и среднее арифметическое столбцов:
 - Rating
 - Reviews

In [20]:
df['rating']

0        4.1
1        3.9
2        4.7
3        4.5
4        4.3
        ... 
10836    4.5
10837    5.0
10838    0.0
10839    4.5
10840    4.5
Name: rating, Length: 10829, dtype: float64

In [21]:
median_rating = df.rating.median()
median_rating

4.2

In [22]:
df['rating'].mean()

3.6231969710961307

In [23]:
df['reviews']

0           159
1           967
2         87510
3        215644
4           967
          ...  
10836        38
10837         4
10838         3
10839       114
10840    398307
Name: reviews, Length: 10829, dtype: object

In [24]:
df['reviews'] = pd.to_numeric(df['reviews'], errors='ignore') # преобразования объекта в число тк был dtype: object

C:\Users\olyao\AppData\Local\Temp\ipykernel_16828\2826382456.py:1: FutureWarning: errors='ignore' is deprecated and will raise in a future version. Use to_numeric without passing `errors` and catch exceptions explicitly instead
  df['reviews'] = pd.to_numeric(df['reviews'], errors='ignore') # преобразования объекта в число тк был dtype: object


In [25]:
df['reviews'].median()

2100.0

In [26]:
df['reviews'].mean().round() # большая разница тк в солбце есть выбросы

444602.0

8. Выведите на экран все уникальные значения категориального столбца Genres

In [27]:
df['genres'].unique()

array(['Art & Design', 'Art & Design;Pretend Play',
       'Art & Design;Creativity', 'Art & Design;Action & Adventure',
       'Auto & Vehicles', 'Beauty', 'Books & Reference', 'Business',
       'Comics', 'Comics;Creativity', 'Communication', 'Dating',
       'Education;Education', 'Education', 'Education;Creativity',
       'Education;Music & Video', 'Education;Action & Adventure',
       'Education;Pretend Play', 'Education;Brain Games', 'Entertainment',
       'Entertainment;Music & Video', 'Entertainment;Brain Games',
       'Entertainment;Creativity', 'Events', 'Finance', 'Food & Drink',
       'Health & Fitness', 'House & Home', 'Libraries & Demo',
       'Lifestyle', 'Lifestyle;Pretend Play',
       'Adventure;Action & Adventure', 'Arcade', 'Casual', 'Card',
       'Casual;Pretend Play', 'Action', 'Strategy', 'Puzzle', 'Sports',
       'Music', 'Word', 'Racing', 'Casual;Creativity',
       'Casual;Action & Adventure', 'Simulation', 'Adventure', 'Board',
       'Trivia', 'Role 

9. Сгруппируйте данные по столбцу `Genres` и посчитайте для каждого жанра средний и медианный рейтинг. Представьте результат в виде нового DataFrame,который будет сохранен в переменную `grouped_df`, где одна колонка будет содержать названия жарнов, а две другие буду содержать в себе средний и медианный рейтинги. Напишите краткий вывод.

In [28]:
grouped_df = df.groupby("genres").agg({"rating": ["mean", "median"]}).round(1) # Группируем данные по столбцу genres и вычисляем средний и медианный рейтинг

In [29]:
grouped_df 

rating       
                                        mean median
genres                                             
Action                                   4.2    4.3
Action;Action & Adventure                4.3    4.3
Adventure                                4.1    4.3
Adventure;Action & Adventure             4.4    4.5
Adventure;Brain Games                    4.6    4.6
...                                      ...    ...
Video Players & Editors                  3.7    4.2
Video Players & Editors;Creativity       4.1    4.1
Video Players & Editors;Music & Video    4.0    4.0
Weather                                  3.9    4.2
Word                                     4.3    4.4

[119 rows x 2 columns]

In [30]:
grouped_df.columns = ["mean_rating", "median_rating"] # Переименовываем колонки

In [31]:
grouped_df["genres"] = grouped_df.index # Добавляем столбец genres

In [32]:
grouped_df = grouped_df[["genres"] + [col for col in grouped_df.columns if col != "genres"]] # Переставляем столбцы, чтобы genres был первой колонкой

In [33]:
grouped_df = grouped_df.reset_index(drop=True) # Удаляем предыдущий индекс и сохраняем изменения
grouped_df.head()

genres  mean_rating  median_rating
0                        Action          4.2            4.3
1     Action;Action & Adventure          4.3            4.3
2                     Adventure          4.1            4.3
3  Adventure;Action & Adventure          4.4            4.5
4         Adventure;Brain Games          4.6            4.6

In [34]:
grouped_df.mean_rating.value_counts(bins=10) # Смотрим средний рейтинг приложений

(3.84, 4.32]       56
(4.32, 4.8]        26
(3.36, 3.84]       25
(2.88, 3.36]        6
(-0.0058, 0.48]     4
(1.44, 1.92]        1
(2.4, 2.88]         1
(0.48, 0.96]        0
(0.96, 1.44]        0
(1.92, 2.4]         0
Name: count, dtype: int64

In [35]:
grouped_df.median_rating.value_counts(bins=10) # Смотрим медиану рейтинга приложений

(3.84, 4.32]       76
(4.32, 4.8]        37
(-0.0058, 0.48]     5
(3.36, 3.84]        1
(0.48, 0.96]        0
(0.96, 1.44]        0
(1.44, 1.92]        0
(1.92, 2.4]         0
(2.4, 2.88]         0
(2.88, 3.36]        0
Name: count, dtype: int64

Среднее и медианное значения рейтинга приложений различаются. Это видно из того, что столбцы mean_rating и median_rating в grouped_df имеют разные распределения:

* Среднее значение: Большинство значений среднего рейтинга находятся в диапазоне (3,84, 4,32], при этом 56 приложений имеют средний рейтинг в этом диапазоне.
* Медианное значение: Большинство значений медианного рейтинга также находятся в диапазоне (3,84, 4,32], но при этом 76 приложений имеют медианный рейтинг в этом диапазоне.

Разница между средним и медианным значениями указывает на то, что некоторые приложения имеют очень высокие или очень низкие рейтинги, что влияет на среднее значение. Медианное значение менее чувствительно к выбросам, поэтому оно может быть более репрезентативным для типичного рейтинга приложений в каждой категории.

Вывод:

Средняя и медиана рейтинга приложений различаются, что указывает на наличие выбросов в данных о рейтингах. Медианное значение может быть более репрезентативным для типичного рейтинга приложений в каждой категории, поскольку оно менее чувствительно к выбросам.

In [36]:
grouped_df.shape

(119, 3)

In [37]:
grouped_df.sort_values(by='mean_rating', ascending=True) 

genres  mean_rating  median_rating
113                 Trivia;Education          0.0            0.0
94            Role Playing;Education          0.0            0.0
10   Art & Design;Action & Adventure          0.0            0.0
20      Books & Reference;Creativity          0.0            0.0
78               Parenting;Education          1.7            0.0
..                               ...          ...            ...
4              Adventure;Brain Games          4.6            4.6
87                  Puzzle;Education          4.6            4.6
64        Health & Fitness;Education          4.7            4.7
18                Board;Pretend Play          4.8            4.8
35                 Comics;Creativity          4.8            4.8

[119 rows x 3 columns]

10. Из получившегося датафрейма выведите на экран только те жанры, чей медианный рейтинг больше 4.5. Напишите краткий вывод.

In [38]:
filtered_df = grouped_df[grouped_df["median_rating"] > 4.5] # Фильтруем датафрейм, оставляя только жанры с медианным рейтингом больше 4.5.
filtered_df # Из всего рейтинга видим что только 9 жанров приложений из 113 попадают под критерий рейтинга более 4.5

genres  mean_rating  median_rating
4          Adventure;Brain Games          4.6            4.6
11       Art & Design;Creativity          4.4            4.7
18            Board;Pretend Play          4.8            4.8
35             Comics;Creativity          4.8            4.8
55      Entertainment;Creativity          4.5            4.6
64    Health & Fitness;Education          4.7            4.7
74           Music;Music & Video          4.5            4.6
87              Puzzle;Education          4.6            4.6
105  Strategy;Action & Adventure          4.6            4.6

11. Выведете кол-во дубликатов в основном DataFrame (df) и удалите их 

In [39]:
df.duplicated().sum()

483

In [40]:
df.drop_duplicates(inplace=True)
df.duplicated().sum()

0

12. Посчитатайте какое количество приложений содержится в каждом жанре (value_counts).В получившемся Series выведите на экран 10 строк с наибольшим количеством приложений отсортированных по убыванию.

In [41]:
genre_counts = df["genres"].value_counts()
genre_counts

genres
Tools                                840
Entertainment                        587
Education                            526
Business                             427
Medical                              408
                                    ... 
Parenting;Brain Games                  1
Travel & Local;Action & Adventure      1
Lifestyle;Pretend Play                 1
Tools;Education                        1
Strategy;Creativity                    1
Name: count, Length: 119, dtype: int64

In [42]:
genre_counts.head(10)

genres
Tools              840
Entertainment      587
Education          526
Business           427
Medical            408
Productivity       407
Personalization    386
Lifestyle          372
Communication      366
Sports             364
Name: count, dtype: int64

# PRO часть (Не обязательна к выполнению, но желательна :))


Измените названия всех столбцов приведя их к нижнему регистру

In [43]:
df.columns # Выводит названия столбцов датафрейма.

Index(['app', 'category', 'rating', 'reviews', 'size', 'installs', 'type',
       'price', 'content_rating', 'genres', 'last_updated', 'current_ver',
       'android_ver'],
      dtype='object')

In [44]:
import re

In [45]:
df.columns = [re.sub('(?<!^)(?=[A-Z])', '_', i, count=1).lower() for i in df]
df.head()

app        category  rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

   reviews  size     installs  type  price content_rating  \
0      159   19M      10,000+  Free    0.0       Everyone   
1      967   14M     500,000+  Free    0.0       Everyone   
2    87510  8.7M   5,000,000+  Free    0.0       Everyone   
3   215644   25M  50,000,000+  Free    0.0           Teen   
4      967  2.8M     100,000+  Free    0.0       Everyone   

                      genres      last_updated         current_ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    android_ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

Отобразите количество строк и столбцов в DataFrame

In [46]:
df.shape # Подсчитываем количкство строк и столбцов

(10346, 13)

Выведите основные статистики по категориальным столбцам. Напишите выводы о них 

In [47]:
df.describe(include='object')

app category                size    installs   type content_rating  \
count    10346    10346               10346       10346  10346          10346   
unique    9648       33                 457          20      2              6   
top     ROBLOX   FAMILY  Varies with device  1,000,000+   Free       Everyone   
freq         9     1939                1525        1487   9584           8372   

       genres    last_updated         current_ver android_ver  
count   10346           10346               10346       10346  
unique    119            1376                2831          33  
top     Tools  August 3, 2018  Varies with device  4.1 and up  
freq      840             298                1301        2379

Выводы:

* app: Существует большое разнообразие приложений в магазине, всего 9660 уникальных приложений.
* category: Самой популярной категорией является FAMILY, на которую приходится 1972 приложения.
* reviews: Количество отзывов сильно варьируется, причем 596 приложений не имеют отзывов.
* size: Размеры приложений также сильно варьируются, причем наиболее распространенным размером является "Varies with device".
* installs: Большинство приложений имеют более 1 миллиона установок.
* type: Подавляющее большинство приложений бесплатны.
* content rating: Большинство приложений имеют рейтинг "Everyone", что указывает на то, что они подходят для всех возрастов.
* genres: Существует большое разнообразие жанров приложений, причем наиболее распространенным жанром является Tools.
* last updated: Приложения регулярно обновляются, при этом большинство из них обновлялось в течение последних нескольких месяцев.
* current ver: Версии приложений также сильно различаются.
* android ver: Большинство приложений требуют Android версии 4.1 и выше.

Эти статистики показывают, что рынок мобильных приложений очень разнообразен и конкурентоспособен. Разработчикам приложений необходимо тщательно выбирать категории, жанры и целевую аудиторию, чтобы добиться успеха в магазине приложений.

Выведите на экран часть датафрейма, который соответствует условию: `приложение из категории искусства и дизайна` И `с рейтингом больше 4.5`

In [48]:
df['category'].unique()

array(['ART_AND_DESIGN', 'AUTO_AND_VEHICLES', 'BEAUTY',
       'BOOKS_AND_REFERENCE', 'BUSINESS', 'COMICS', 'COMMUNICATION',
       'DATING', 'EDUCATION', 'ENTERTAINMENT', 'EVENTS', 'FINANCE',
       'FOOD_AND_DRINK', 'HEALTH_AND_FITNESS', 'HOUSE_AND_HOME',
       'LIBRARIES_AND_DEMO', 'LIFESTYLE', 'GAME', 'FAMILY', 'MEDICAL',
       'SOCIAL', 'SHOPPING', 'PHOTOGRAPHY', 'SPORTS', 'TRAVEL_AND_LOCAL',
       'TOOLS', 'PERSONALIZATION', 'PRODUCTIVITY', 'PARENTING', 'WEATHER',
       'VIDEO_PLAYERS', 'NEWS_AND_MAGAZINES', 'MAPS_AND_NAVIGATION'],
      dtype=object)

In [49]:
art_and_design_df = df[(df['category'] == 'ART_AND_DESIGN') & (df['rating'] > 4.5)]
art_and_design_df.sort_values('rating', ascending=False)

app        category  \
8888            Spring flowers theme couleurs d t space  ART_AND_DESIGN   
6930               Fantasy theme dark bw black building  ART_AND_DESIGN   
5254                            AJ Styles HD Wallpapers  ART_AND_DESIGN   
4759  X Launcher Pro - IOS Style Theme & Control Center  ART_AND_DESIGN   
25                           Harley Quinn wallpapers HD  ART_AND_DESIGN   
3982                                  Cardi B Wallpaper  ART_AND_DESIGN   
9                         Kids Paint Free - Drawing Fun  ART_AND_DESIGN   
4764     X Launcher Prime: With OS Style Theme & No Ads  ART_AND_DESIGN   
4755  X Launcher Pro: PhoneX Theme, OS11 Control Center  ART_AND_DESIGN   
4749  X Launcher: With OS11 Style Theme & Control Ce...  ART_AND_DESIGN   
45    Canva: Poster, banner, card maker & graphic de...  ART_AND_DESIGN   
2     U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN   
36       UNICORN - Color By Number & Pixel Art Coloring  ART_AND_DESIGN   
34                                      I Creative Idea  ART_AND_DESIGN   
26                        Colorfit - Drawing & Coloring  ART_AND_DESIGN   
24                          HD Mickey Minnie Wallpapers  ART_AND_DESIGN   
22              Superheroes Wallpapers | 4K Backgrounds  ART_AND_DESIGN   
16         Photo Designer - Write your name with shapes  ART_AND_DESIGN   
38                       PIP Camera - PIP Collage Maker  ART_AND_DESIGN   
46    Install images with music to make video withou...  ART_AND_DESIGN   
19                                         ibis Paint X  ART_AND_DESIGN   
13                                Mandala Coloring Book  ART_AND_DESIGN   

      rating  reviews  size     installs  type  price content_rating  \
8888     5.0        1  2.9M         100+  Free   0.00       Everyone   
6930     4.8       41  1.9M       5,000+  Free   0.00       Everyone   
5254     4.8       48   25M       5,000+  Free   0.00       Everyone   
4759     4.8     1216  8.6M      10,000+  Paid   1.99       Everyone   
25       4.8      192  6.0M      10,000+  Free   0.00       Everyone   
3982     4.8      253  3.7M      50,000+  Free   0.00       Everyone   
9        4.7      121  3.1M      10,000+  Free   0.00       Everyone   
4764     4.7      149  3.5M       1,000+  Paid   1.99       Everyone   
4755     4.7      801  3.5M       5,000+  Paid   1.99       Everyone   
4749     4.7     5754  4.4M     100,000+  Free   0.00       Everyone   
45       4.7   174531   24M  10,000,000+  Free   0.00       Everyone   
2        4.7    87510  8.7M   5,000,000+  Free   0.00       Everyone   
36       4.7     8145   24M     500,000+  Free   0.00       Everyone   
34       4.7      353  4.2M      10,000+  Free   0.00           Teen   
26       4.7    20260   25M     500,000+  Free   0.00       Everyone   
24       4.7      118   23M      50,000+  Free   0.00       Everyone   
22       4.7     7699  4.2M     500,000+  Free   0.00   Everyone 10+   
16       4.7     3632  5.5M     500,000+  Free   0.00       Everyone   
38       4.7      158   11M      10,000+  Free   0.00       Everyone   
46       4.6     1070   26M     100,000+  Free   0.00       Everyone   
19       4.6   224399   31M  10,000,000+  Free   0.00       Everyone   
13       4.6     4326   21M     100,000+  Free   0.00       Everyone   

                       genres       last_updated current_ver   android_ver  
8888             Art & Design     April 18, 2018       1.0.2    4.0 and up  
6930             Art & Design     March 20, 2018       1.0.2    4.0 and up  
5254             Art & Design     August 4, 2018       1.1.1    4.1 and up  
4759             Art & Design      June 26, 2018       1.0.0    4.1 and up  
25               Art & Design     April 25, 2018         1.5    3.0 and up  
3982             Art & Design   November 1, 2017       1.0.0    4.0 and up  
9     Art & Design;Creativity       July 3, 2018         2.8  4.0.3 and up  
4764             Art & Design      July 30, 2018     

Посчитатайте какое количество игр содержится в рейтинге, используя параметр разбиения (bins) равным 100. Выведите на экран топ-10 игр по количеству в рейтинге. Напишите краткий вывод.
То есть сначала отфильтровываете лишь игры из всего датафрейма, после чего применяете value_counts к столбцу с рейтингом используя параметр, который указан выше. И выводите первые 10 строк отсортированных по убыванию.

In [50]:
games_df = df[df['category'] == 'GAME']
games_df.shape[0]

1121

In [51]:
rating_counts = games_df['rating'].value_counts(bins=100)
rating_counts

(4.45, 4.5]    169
(4.35, 4.4]    162
(4.25, 4.3]    148
(4.15, 4.2]    115
(4.55, 4.6]    113
              ... 
(2.05, 2.1]      0
(2.0, 2.05]      0
(1.95, 2.0]      0
(1.9, 1.95]      0
(2.5, 2.55]      0
Name: count, Length: 100, dtype: int64

In [52]:
rating_counts.head(10)

(4.45, 4.5]       169
(4.35, 4.4]       162
(4.25, 4.3]       148
(4.15, 4.2]       115
(4.55, 4.6]       113
(4.05, 4.1]        72
(4.65, 4.7]        65
(3.95, 4.0]        55
(-0.006, 0.05]     47
(3.75, 3.8]        41
Name: count, dtype: int64

In [53]:
rating_counts.head(10).sum()

987

Создайте функцию , которая будет применена к столбцу рейтинг методом `apply()`.
Функция должна принимать строку с рейтингом и возвращать:
 - High rating, если рейтинг больше или равен 4.5
 - Middle rating, если рейтинг между 3.8 и 4.5
 - Low rating, если рейтинг меньше 3.8 
Результат работы функции запишите в новый столбец с названием `categorical_rating`

In [54]:
def rating_category(rating):
    rating = float(rating)
    if rating >= 4.5:
        return "High rating"
    elif rating >= 3.8:
        return "Middle rating"
    else:
        return "Low rating"
df['categorical_rating'] = df['rating'].apply(rating_category)
df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating  reviews                size     installs  type  price  \
0         4.1      159                 19M      10,000+  Free    0.0   
1         3.9      967                 14M     500,000+  Free    0.0   
2         4.7    87510                8.7M   5,000,000+  Free    0.0   
3         4.5   215644                 25M  50,000,000+  Free    0.0   
4         4.3      967                2.8M     100,000+  Free    0.0   
...       ...      ...                 ...          ...   ...    ...   
10836     4.5       38                 53M       5,000+  Free    0.0   
10837     5.0        4                3.6M         100+  Free    0.0   
10838     0.0        3                9.5M       1,000+  Free    0.0   
10839     4.5      114  Varies with device       1,000+  Free    0.0   
10840     4.5   398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver categorical_rating  
0                   1.0.0        4.0.3 and up      Middle rating  
1                   2.0.0        4.0.3 and up      Middle rating  
2                   1.2.4        4.0.3 and up        High rating  
3      Varies with device          4.2 and up        High rating  
4                     1.1          4.4 and up      Middle rating  
...                   ...                 ...                ...  
10836                1.48          4.1 and up        High rating  
10837                 1.0          4.1 and up        High rating  
10838                 1.0          2.2 and up         Low rating  
10839  Varies with device  Varies with device        High rating  
10840  Varies with device  Varies with device        High rating  

[10346 rows x 14 columns]

Выведите кол-во уникальных значений в столбцах отсортированных по убыванию на экран:

In [55]:
df.nunique().sort_values(ascending=False)

app                   9648
reviews               5999
current_ver           2831
last_updated          1376
size                   457
genres                 119
price                   92
rating                  40
category                33
android_ver             33
installs                20
content_rating           6
categorical_rating       3
type                     2
dtype: int64

__Super-Hard задания__

Решите задачу, где мы создавали свою собственную функцию, но вместо функции `def` используйте `лямбда-функцию` внутри `apply()`

In [56]:
df['categorical_rating'] = df['rating'].apply(lambda x: "High rating" if x >= 4.5 else ("Middle rating" if x >= 3.8 else "Low rating"))
df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating  reviews                size     installs  type  price  \
0         4.1      159                 19M      10,000+  Free    0.0   
1         3.9      967                 14M     500,000+  Free    0.0   
2         4.7    87510                8.7M   5,000,000+  Free    0.0   
3         4.5   215644                 25M  50,000,000+  Free    0.0   
4         4.3      967                2.8M     100,000+  Free    0.0   
...       ...      ...                 ...          ...   ...    ...   
10836     4.5       38                 53M       5,000+  Free    0.0   
10837     5.0        4                3.6M         100+  Free    0.0   
10838     0.0        3                9.5M       1,000+  Free    0.0   
10839     4.5      114  Varies with device       1,000+  Free    0.0   
10840     4.5   398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver categorical_rating  
0                   1.0.0        4.0.3 and up      Middle rating  
1                   2.0.0        4.0.3 and up      Middle rating  
2                   1.2.4        4.0.3 and up        High rating  
3      Varies with device          4.2 and up        High rating  
4                     1.1          4.4 and up      Middle rating  
...                   ...                 ...                ...  
10836                1.48          4.1 and up        High rating  
10837                 1.0          4.1 and up        High rating  
10838                 1.0          2.2 and up         Low rating  
10839  Varies with device  Varies with device        High rating  
10840  Varies with device  Varies with device        High rating  

[10346 rows x 14 columns]

Изучите функцию `transform` в документации Pandas. Примените её к датафрейму, который будет заранее сгруппирован по столбцу `category`. С помощью фукнции `transform` найдите средний `rating` по каждой категории и запишите в столбец с названием `mean_cat_rating`. Кол-во строк в получившемся новом столбец должно быть таким же, как и в исходном датафрейме

In [57]:
# grouped_df = df.groupby('category') # Группируем датафрейм по столбцу category
# mean_cat_ratings = grouped_df['rating'].transform('mean').round(2) # Находим средний рейтинг по каждой категории
# df['mean_cat_rating'] = mean_cat_ratings # Добавляем столбец со средними рейтингами в исходный датафрейм

df['mean_cat_rating'] = df.groupby('categorical_rating')['rating'].transform('mean').round(2)
df.sample(5) # Возвращаем случайную выборку строк датафрейма.  

app        category  \
2628   Nextdoor - Local neighborhood news & classifieds          SOCIAL   
4268                                           J.ME Fly          FAMILY   
3517                                      File Explorer    PRODUCTIVITY   
10594                        FL Racing Manager 2018 Pro          SPORTS   
12                       Tattoo Name On My Photo Editor  ART_AND_DESIGN   

       rating  reviews  size     installs  type  price content_rating  \
2628      4.3    51504   20M   5,000,000+  Free   0.00           Teen   
4268      3.9       28   98M       1,000+  Free   0.00       Everyone   
3517      4.1      979  720k     100,000+  Free   0.00       Everyone   
10594     4.3      340   15M       5,000+  Paid   1.99       Everyone   
12        4.2    44829   20M  10,000,000+  Free   0.00           Teen   

              genres        last_updated current_ver android_ver  \
2628          Social       July 25, 2018        2.69  5.0 and up   
4268   Entertainment        May 31, 2018       0.4.4  4.4 and up   
3517    Productivity  September 30, 2013       2.1.3  2.2 and up   
10594         Sports      March 17, 2018        1.18  3.0 and up   
12      Art & Design       April 2, 2018         3.8  4.1 and up   

      categorical_rating  mean_cat_rating  
2628       Middle rating             4.18  
4268       Middle rating             4.18  
3517       Middle rating             4.18  
10594      Middle rating             4.18  
12         Middle rating             4.18

Создайте функцию , которая будет применена к столбцу `last updated` и возвращать из него число месяца, если число месяца вернуть из строки невозможно, то верните вместо этого строку с значением `miss_date`.\
Запишите полученный столбец в датафрейм , название столбцу дайте `day_of_update`.

Если вы столкнётесь с некоторой ошибкой , которую вам выдаст Pandas, то изучите конструкцию Python `try` \ `except`.\
Вот псевдокод , который поможет лучше разобраться с этой конструкцией:

In [58]:
df = df.rename(columns={'last updated': 'last_updated'}) # Переименновываем колонку

In [60]:
def day_of_update(date_string):
    try: # Помещаете то, что должно выполниться внутрь try, если же внутри try появляется ошибка, то условие переключается на оператор except и выполняется всё то, что находится внутри except.
        datetime_object = pd.to_datetime(date_string) # Преобразуем строку с датой в объект datetime
        day_of_update = datetime_object.month # Извлекаем число месяца из объекта datetime
        return day_of_update # Возвращаем число месяца
    except:
        return 'miss_date' # Если преобразовать строку с датой в объект datetime невозможно, возвращаем строку с значением miss_date
    
df['day_of_update'] = df['last_updated'].apply(day_of_update)
df

app             category  \
0         Photo Editor & Candy Camera & Grid & ScrapBook       ART_AND_DESIGN   
1                                    Coloring book moana       ART_AND_DESIGN   
2      U Launcher Lite – FREE Live Cool Themes, Hide ...       ART_AND_DESIGN   
3                                  Sketch - Draw & Paint       ART_AND_DESIGN   
4                  Pixel Draw - Number Art Coloring Book       ART_AND_DESIGN   
...                                                  ...                  ...   
10836                                   Sya9a Maroc - FR               FAMILY   
10837                   Fr. Mike Schmitz Audio Teachings               FAMILY   
10838                             Parkinson Exercices FR              MEDICAL   
10839                      The SCP Foundation DB fr nn5n  BOOKS_AND_REFERENCE   
10840      iHoroscope - 2018 Daily Horoscope & Astrology            LIFESTYLE   

       rating  reviews                size     installs  type  price  \
0         4.1      159                 19M      10,000+  Free    0.0   
1         3.9      967                 14M     500,000+  Free    0.0   
2         4.7    87510                8.7M   5,000,000+  Free    0.0   
3         4.5   215644                 25M  50,000,000+  Free    0.0   
4         4.3      967                2.8M     100,000+  Free    0.0   
...       ...      ...                 ...          ...   ...    ...   
10836     4.5       38                 53M       5,000+  Free    0.0   
10837     5.0        4                3.6M         100+  Free    0.0   
10838     0.0        3                9.5M       1,000+  Free    0.0   
10839     4.5      114  Varies with device       1,000+  Free    0.0   
10840     4.5   398307                 19M  10,000,000+  Free    0.0   

      content_rating                     genres      last_updated  \
0           Everyone               Art & Design   January 7, 2018   
1           Everyone  Art & Design;Pretend Play  January 15, 2018   
2           Everyone               Art & Design    August 1, 2018   
3               Teen               Art & Design      June 8, 2018   
4           Everyone    Art & Design;Creativity     June 20, 2018   
...              ...                        ...               ...   
10836       Everyone                  Education     July 25, 2017   
10837       Everyone                  Education      July 6, 2018   
10838       Everyone                    Medical  January 20, 2017   
10839     Mature 17+          Books & Reference  January 19, 2015   
10840       Everyone                  Lifestyle     July 25, 2018   

              current_ver         android_ver categorical_rating  \
0                   1.0.0        4.0.3 and up      Middle rating   
1                   2.0.0        4.0.3 and up      Middle rating   
2                   1.2.4        4.0.3 and up        High rating   
3      Varies with device          4.2 and up        High rating   
4                     1.1          4.4 and up      Middle rating   
...                   ...                 ...                ...   
10836                1.48          4.1 and up        High rating   
10837                 1.0          4.1 and up        High rating   
10838                 1.0          2.2 and up         Low rating   
10839  Varies with device  Varies with device        High rating   
10840  Varies with device  Varies with device        High rating   

       mean_cat_rating  day_of_update  
0                 4.18              1  
1                 4.18              1  
2                 4.65              8  
3                 4.65              6  
4                 4.18              6  
...                ...            ...  
10836             4.65              7  
10837             4.65              7  
10838             1.50              1  
10839             4.65              1  
10840             4.65              7  

[10346 rows x 16 columns]